### PR TITLE
config: document interface and mark types as public

### PR DESCRIPTION
**Pull request by Rugvip**
🧹

### CODE REVIEW

File: packages/config/src/types.ts [hunk 1]
```diff
@@ -37,28 +81,90 @@ export type Config = {
...
 
+  /**
+   * Same as `getOptional`, but will throw an error if there's no value for the given key.
+   */
   get<T = JsonValue>(key?: string): T;
+
+  /**
+   * Read out all configuration data for the given key.
+   *
+   * Usage of this method should be avoided as using the typed alternatives provide
```
Comment (freben) [inline packages/config/src/types.ts:102]:
```suggestion
   * Usage of this method should be avoided as using the typed alternatives provides
```

File: packages/config/src/types.ts [hunk 2]
```diff
@@ -37,28 +81,90 @@ export type Config = {
...
+   * much better error reporting. The main use-case of this method is to determine
+   * the type on a configuration value in the case where there are multiple possible
```
Comment (freben) [inline packages/config/src/types.ts:104]:
```suggestion
   * the type of a configuration value in the case where there are multiple possible
```

**Review by freben — APPROVED** [review]

File: packages/config/src/types.ts (hunk 1 above)
Comment (Rugvip) [reply packages/config/src/types.ts:102]: But there are so many alternatives

Comment (freben) [reply packages/config/src/types.ts:102]: But only one using. (It really should be singular :))

Comment (Rugvip) [reply packages/config/src/types.ts:102]: Wat, no

Comment (freben) [reply packages/config/src/types.ts:102]: @timbonicus

Comment (freben) [reply packages/config/src/types.ts:102]: The count of things doesn't come into play at all. Try putting "leads to" there instead to get a feel for it.

"Using X provides better Y"

It's the usage that provides, not the X.

Comment (Rugvip) [reply packages/config/src/types.ts]: Repetition and discussion gone x)

Comment (timbonicus) [reply packages/config/src/types.ts]: Edit LGTM!

**Pull request merged**